docs/SPEC-composition.md

```diff
@@ -0,0 +1,173 @@
+Flux Composition
+==
+
+The composition capabilities of the flux LSP server enable clients to create a query builder experience without deep knowledge of the flux internals. Composition will create a "Composition-owned" statement in a given flux file, and will be able to add/remove filter calls to that statement. The composition rules apply as follows:
+
+  - Composition must never generate a query that cannot be executed.
+  - Composition should never generate a query that a user with sufficient flux knowledge would never write, e.g. `filter(fn: (r) => r.foo == "bar") |> filter(fn: (r) => r.foo == "bar")`
+  - Composition should never generate a query that results in immediate diagnostic messages.
+  - In the event of conflicts or logical ambiguity, Composition should default to erroring rather than introducing complex assumption logic to attempt recovery.
+
+Given these rules, this document will not attempt to document the resulting flux code generated from each of these requests, as the resulting flux could change as requirements evolve and the corresponding flux would become outdated.
+
+Composition Initialize
+--
+
+The composition initialize request is sent from the client to the server to identify a file that will be used to compose queries. This call must be made before any other Composition requests, otherwise those requests will error. It will result in a `workspace/applyEdit` request from the server to the client with the associated changes to the file. The request can also be used to reset the composition query state by calling it again, and can be considered idempotent.
+
+*Request*
+  - method: `fluxComposition/initialize`
+  - params: `CompositionInitializeParams`
+
+```
+interface CompositionInitializeParams {
+    /*
+     * The text document.
+     */
+    text_document: TextDocumentIdentifier;
+
+    /*
+     * The bucket to initialize the composition query.
+     */
+    bucket: string;
+
+    /*
+     * The measurement to initialize the composition query.
+     */
+    measurement?: string;
+}
```

Comment:
Should we also include a timeRange in the `fluxComposition/initialize`? Or does it always hardcode to the params/v properties?

Reply:
I think, for now, it makes sense to hardcode it, if for no other reason than we don't really have a nice way to specify timeranges. The LSP already has an established pattern for adding in parameters after the fact, so we're not painted into a corner there.